clang/include/clang/Basic/Builtins.td

```diff
@@ -4894,7 +4894,7 @@ def HLSLDotProduct : LangBuiltin<"HLSL_LANG"> {
 def HLSLDot2Add : LangBuiltin<"HLSL_LANG"> {
   let Spellings = ["__builtin_hlsl_dot2add"];
   let Attributes = [NoThrow, Const, CustomTypeChecking];
```

Comment:
now that you defined the types you don't need `CustomTypeChecking`.

Reply:
Removing this makes Codegen unhappy. It is not able to implicitly convert `half` to `_Float16_t`. I think we need to find a way to use `half` here.

Reply:
Not sure what you mean. Are you saying you are seeing a crash when you run `clang/test/CodeGenHLSL/builtins/dot2add.hlsl` or does the test case not pass? Is it unhappy with the cases that don't use `-fnative-half-type` or does that flag not affect the result?

My theory is maybe the fact that HLSL treats halfs as floats without `-fnative-half-type` might be casuing a problem but don't feel like there is enoug information.

Reply:
`clang.exe` is not able to compile `clang/test/CodeGenHLSL/builtins/dot2add.hlsl` even after passing the flag `-fnative-half-type` saying 

`error: passing 'vector<_Float16, 2>' (vector of 2 '_Float16' values) to parameter of incompatible type '__attribute__((__vector_size__(2 * sizeof(half)))) half' (vector of 2 'half' values)`.

Reply:
It looks like `_fp16` is the better supported float 16 type  used heaviliy in the Nvidia gpu builtins. You should find that this error goes away if you switch to it. That said I've seen comments that indicate it has something to do with OpenCL so  maybe ask @llvm-beanz if it makes sense for us to use `_fp16`.